### ISSUE TITLE
anyio.BrokenResourceError exception not caught ?

### ISSUE THREAD

**Issue by devl00p** [encode/httpcore#382]
Hello,

I spotted several cases where an anyio exceptions made my program crash.

The error that is mentionned every time is `anyio.BrokenResourceError` but the traceback may change.

I collected the following ones:

```
Future exception was never retrieved
future: <Future finished exception=BrokenResourceError()>
Traceback (most recent call last):
  File "/usr/lib64/python3.8/asyncio/selector_events.py", line 848, in _read_ready__data_received
    data = self._sock.recv(self.max_size)
ConnectionResetError: [Errno 104] Connection reset by peer

The above exception was the direct cause of the following exception:

anyio.BrokenResourceError
```

```
Future exception was never retrieved
future: <Future finished exception=BrokenResourceError()>
Traceback (most recent call last):
  File "/usr/lib64/python3.8/asyncio/selector_events.py", line 848, in _read_ready__data_received
    data = self._sock.recv(self.max_size)
TimeoutError: [Errno 110] Connection timed out

The above exception was the direct cause of the following exception:

anyio.BrokenResourceError
```

```
Future exception was never retrieved
future: <Future finished exception=BrokenResourceError()>
Traceback (most recent call last):
  File "/usr/lib64/python3.8/asyncio/selector_events.py", line 848, in _read_ready__data_received
    data = self._sock.recv(self.max_size)
ConnectionResetError: [Errno 104] Connection reset by peer

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/devloop/.local/share/virtualenvs/wapiti-p7I6n6KS/lib/python3.8/site-packages/httpcore/_backends/anyio.py", line 60, in read
    return await self.stream.receive(n)
  File "/home/devloop/.local/share/virtualenvs/wapiti-p7I6n6KS/lib/python3.8/site-packages/anyio/_backends/_asyncio.py", line 1093, in receive
    raise self._protocol.exception
anyio.BrokenResourceError
```

What I was expecting: an `httpx` exception instead (RequestError).

Used libraries and versions:

```
aiocache==0.11.1
anyio==3.2.1
async-timeout==3.0.1
httpcore==0.13.6
httpx==0.18.2
```

```Python 3.8.10```

OS: Linux (openSUSE Tumbleweed)

I managed to reproduce with the following script. Unfortunately it doesn't happen for every website.

```python
import asyncio
import signal
import sys

import httpx

MAX_TASKS = 30

stop_event = asyncio.Event()


def stop_attack_process():
    global stop_event
    print("Stopping tasks")
    stop_event.set()


class Buster:
    def __init__(self, root_url: str, payloads_file: str, event: asyncio.Event):
        self._client = httpx.AsyncClient(timeout=10)
        self._root_url = root_url
        self._payloads_file = payloads_file
        self._stop_event = event
        self.network_errors = 0

    async def close(self):
        await self._client.aclose()

    async def check_url(self, url):
        response = await self._client.get(url)
        if response.status_code == 200:
            return True, url
        return False, url

    async def brute(self):
        tasks = set()
        pending_count = 0
        payload_iterator = open(self._payloads_file, errors="ignore")

        while True:
            if pending_count < MAX_TASKS and not self._stop_event.is_set():
                try:
                    candidate = next(payload_iterator)
                except StopIteration:
                    pass
                else:
                    candidate = candidate.strip()
                    if not candidate:
                        continue

                    url = self._root_url + candidate
                    task = asyncio.create_task(self.check_url(url))
                    tasks.add(task)

            if not tasks:
                break

            done_tasks, pending_tasks = await asyncio.wait(
                tasks,
                timeout=0.01,
                return_when=asyncio.FIRST_COMPLETED
            )
            pending_count = len(pending_tasks)
            for task in done_tasks:
                try:
                    result, url = await task
                except httpx.RequestError:
                    self.network_errors += 1
                else:
                    if result:
                        print(f"Found {url}")
                tasks.remove(task)

            if self._stop_event.is_set():
                print("pending tasks:", pending_count)
                for task in pending_tasks:
                    task.cancel()
                    tasks.remove(task)


async def main(root_url: str):
    global stop_event
    filename = "busterPayloads.txt"
    loop = asyncio.get_event_loop()
    loop.add_signal_handler(signal.SIGINT, stop_attack_process)
    buster = Buster(root_url, filename, stop_event)
    await buster.brute()
    loop.remove_signal_handler(signal.SIGINT)
    await buster.close()

if __name__ == "__main__":
    asyncio.run(main(sys.argv[1]))
```

**Comment by florimondmanca** [comment]
Hola,

Just a thought: seems like an exception we ought to be watching in the anyio backend in HTTPCore. Resolution would involve adding an extra catch case there, I believe. :)

**Comment by kangzhang** [comment]
We ran into similar regression when we upgrade to newer version of httpcore which has anyio as the backend. See #393 for stacktrace. 

One fix I tried on application side was to make sure all http responses are properly closed. I have not seen the crash since, but may just be a coincidence (it only has been 12 hours). 

@florimondmanca - did you mean we should add a catch for it at https://github.com/encode/httpcore/blame/master/httpcore/_backends/anyio.py#L60 ? There seems already a catch there which is supposed to convert the exception into a `ReadException`. I don't fully grasp the logic there.

**Comment by kangzhang** [comment]
Closing http response didn't help. We still get this error message printed out at the end of our eventloop. 

Here is a minimal repro I constructed: 

```
import httpx
client = httpx.AsyncClient()


async def hello():
    try:
        response = await client.get("https://socketsite.com/feed/rss")
    except Exception as e:
        pass
    print("finished")

import asyncio
asyncio.run(hello())
```

It outputs

```
finished
Future exception was never retrieved
future: <Future finished exception=BrokenResourceError()>
Traceback (most recent call last):
  File "/usr/lib/python3.8/asyncio/selector_events.py", line 848, in _read_ready__data_received
    data = self._sock.recv(self.max_size)
ConnectionResetError: [Errno 104] Connection reset by peer

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/kang/envs/py38/lib/python3.8/site-packages/httpcore/_exceptions.py", line 8, in map_exceptions
    yield
  File "/home/kang/envs/py38/lib/python3.8/site-packages/httpcore/_backends/anyio.py", line 154, in open_tcp_stream
    stream = await TLSStream.wrap(
  File "/home/kang/envs/py38/lib/python3.8/site-packages/anyio/streams/tls.py", line 94, in wrap
    await wrapper._call_sslobject_method(ssl_object.do_handshake)
  File "/home/kang/envs/py38/lib/python3.8/site-packages/anyio/streams/tls.py", line 109, in _call_sslobject_method
    data = await self.transport_stream.receive()
  File "/home/kang/envs/py38/lib/python3.8/site-packages/anyio/_backends/_asyncio.py", line 1095, in receive
    raise self._protocol.exception
anyio.BrokenResourceError
/home/kang/envs/py38/lib/python3.8/site-packages/httpx/_client.py:1978: UserWarning: Unclosed <httpx.AsyncClient object at 0x7f4fa0928dc0>. See https://www.python-httpx.org/async/#opening-and-closing-clients for details.
```

It seems a uncaught exception from the framework cached somewhere. It's similar to the case mentioned here: https://quantlane.com/blog/ensure-asyncio-task-exceptions-get-logged/

Could someone with more experience of the stack take a look?

**Comment by iwoloschin** [comment]
I have been seeing similar issues on python3.9. It seems to be more prevalent if I'm doing a lot of concurrent requests, though I don't have much detail beyond "more than 10, less than a couple hundred". I'm not certain that high amount of concurrency is actually related to the issue though, I happened to add more concurrency to my project around the same time I upgraded to httpx 0.18/httpcore 0.13.

**Comment by kangzhang** [comment]
I'm taking another look at this bug today and manage to construct a minimal script that repro the issue. It reliably repros the issue on my end.

```
import socket
import threading

import anyio
import httpx

HOST = "localhost"
PORT = 8833


ssck = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
ssck.bind((HOST, PORT))
ssck.listen(1)


def server():
    while True:
        sck, addr = ssck.accept()
        raise Exception("server crashed")


thread = threading.Thread(target=server)
thread.start()
client = httpx.AsyncClient()


async def hello():
    try:
        await client.get(f"http://{HOST}:{PORT}")
    except Exception as e:
        print(e)
    await client.aclose()
    print("finished")


anyio.run(hello)
```

The next step is to look into httpcore's code to see how to patch it. I would love some help on this if the team has time to take a look.

**Comment by kangzhang** [comment]
Looking thru the code and I'm wondering if we should close the stream at here: https://github.com/encode/httpcore/blame/master/httpcore/_backends/anyio.py#L65

Basically adding a line:

`await self.stream.aclose()`

**Comment by kangzhang** [comment]
Did more poking around and this smells like an anyio bug maybe? (or because trio doesn't report the error?) The error is gone once I switch to trio.

```
anyio.run(hello, backend='trio')
```

My hunch is that some of these anyio primitives like `fail_after` did not play cleanly with asyncio code, which caused asyncio to think this exception was unhandled. 

@agronholm I saw that you added anyio support in httpcore, I'm wondering if you may weight in here, thank you!

**Comment by agronholm** [comment]
I can repro the error with your sample script. I'll try to figure out what is going on here.

**Comment by agronholm** [comment]
First notes: that `BrokenResourceError` is indeed supposed to be raised, but the big question is: why is it not handled? That's what I'm going to find out next.

**Comment by kangzhang** [comment]
Thanks for taking a look! Yeah, the exception was handled when the request was executed (the `print(e)` statement was hit during my test), but asyncio seem to disagree. (The error message was printed after "Finished" and everything is GCed).

**Comment by agronholm** [comment]
> Thanks for taking a look! Yeah, the exception was handled when the request was executed (the print(e) statement was hit during my test), but asyncio seem to disagree. (The error message was printed after "Finished" and everything is GCed).

This happens when an exception leaks out of an asynchronous callback (being a task or a low level callback). The exception is raised in an another task so you cannot possibly catch it with your own `except` clause.

**Comment by agronholm** [comment]
The odd thing here is that with the trio backend, `SocketStream.read()` never even gets called. The end result would surely be a `BrokenResourceError` on trio if that method was ever called`. Next I will figure out the cause for this discrepancy.

**Comment by agronholm** [comment]
Ok, it seems that on trio, httpcore doesn't use its `anyio` backend but the `trio` backend instead (which obviously never raises any AnyIO exceptions). This little fact had escaped my attention and explains why it didn't trip my debug lines when running on trio.

Both backends raise `httpcore.ReadError` when trying to read from the socket which is correct. I'm now trying to find out if there is some leftover task that somehow leaked the `BrokenResourceError` on asyncio.

**Comment by agronholm** [comment]
Ok, I finally figured out what's happening here. The `Future` that never got its exception retrieved is `StreamProtocol.write_future` from AnyIO's asyncio backend.

I managed to factor out httpx from your sample script, simplifying it into this:
```python3
import socket
import threading
import sys

import anyio

HOST = "localhost"
PORT = 8833


ssck = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
ssck.bind((HOST, PORT))
ssck.listen(1)


def server():
    sck, addr = ssck.accept()
    del sck  # required to trigger the problem!


thread = threading.Thread(target=server, daemon=True)
thread.start()


async def hello():
    stream = await anyio.connect_tcp(HOST, PORT)
    await stream.send(b'GET')
    try:
        await stream.receive()
    except Exception as e:
        print(e)

    print("finished")


anyio.run(hello)
```
What happens here is the combination of the following:

1. The client **writes** into the socket (skipping will cause the error not to be triggered)
2. The server **aborts the connection** (just closing the accepted socket will cause the error not to be triggered)
3. The `StreamProtocol.connection_lost()` callback gets called with `ConnectionResetError`
4. The callback resolves its `write_future` using `BrokenResourceError`, with that `ConnectionResetError` as the cause
5. The client tries to read from the socket, causing a **new** `BrokenResourceError` to be raised
6. Since nobody tries to write to the stream any more, `write_future` never passes on the first `BrokenResourceError`, hence the warning

While this causes an annoying warning to be printed, it doesn't really break anything. Nevertheless, this is an AnyIO bug that should be fixed. Another oddity is that, when writing a test based on this script, I got `EndOfStream` from the asyncio backend and `BrokenResourceError` from trio. If I place `await sleep(0.1)` before the `send()`, all backends raise `EndOfStream`. Sigh :weary:

**Comment by kangzhang** [comment]
Whoa, thanks for the detailed walk through Alex! 

I was confused by the exception since I saw that we caught it in read(). It now makes sense - it's an exception thrown at a different time. 

And yes our production code runs fine. It's just our crash monitoring (Sentry) is unhappy with this std err output and I'm reluctant to blindly turn it off. 

Looking forward to the fix and thanks again for the help!

**Comment by agronholm** [comment]
If you know how to reliably produce a `ConnectionResetError`, I would appreciate the information.

**Comment by kangzhang** [comment]
Just saw this! - I found the example by googling how `ConnectionReset` error happen and I saw some example buggy code that could trigger it. (and I trimmed it down)

May you have a PR / branch with unit test that I can play with? I'm wondering if it would be more reliable if we block the other thread using semaphone and control the timing more precisely. 

I also checked our logging on this issue, 95% of them are `ConnectionResetError` and 5% of them are `BrokenPipe`. (not sure if it is useful information to you in locating the root cause)

**Comment by agronholm** [comment]
I wrote this test (in `test_sockets.py`, under `TestTCPStream`):
```python3
    async def test_server_crash_after_write(self, family: AnyIPAddressFamily) -> None:
        """
        Tests that there won't be any leftover Futures that don't get their exceptions retrieved.

        See https://github.com/encode/httpcore/issues/382 for details.

        """
        def serve() -> None:
            sock, addr = server_sock.accept()
            sock.close()
            del sock

        server_sock = socket.socket(family, socket.SOCK_STREAM)
        server_sock.settimeout(1)
        server_sock.bind(('localhost', 0))
        server_sock.listen()
        thread = Thread(target=serve, daemon=True)
        thread.start()
        server_addr = server_sock.getsockname()[:2]
        async with await connect_tcp(*server_addr) as stream:
            await sleep(0.1)
            await stream.send(b'hello')
            with pytest.raises(BrokenResourceError):
                await stream.receive()

        thread.join()
```

**Comment by kangzhang** [comment]
Nice! I played with it a bit and following code seems to reliably repro the issue on my box. Main changes:
1) Switched from a Thread to a Process to add back the Exception. (not sure if necessary)
2) Removed `pytest.raise` so that I can see all exceptions (this should be reverted)
3) Added `sleep(0.5)` to server process, so that we wait for the client to connect. (We could potentially change this to Thread + Semaphore (throw server exception only after client TCP connect))
```
    async def test_server_crash_after_write(self, family: AnyIPAddressFamily) -> None:
        """
        Tests that there won't be any leftover Futures that don't get their exceptions retrieved.

        See https://github.com/encode/httpcore/issues/382 for details.

        """
        from multiprocessing import Process
        import time
        def serve() -> None:
            time.sleep(0.5)
            sock, addr = server_sock.accept()
            raise Exception("server crashed")

        server_sock = socket.socket(family, socket.SOCK_STREAM)
        # server_sock.settimeout(1)
        server_sock.bind(('localhost', 0))
        server_sock.listen()
        process = Process(target=serve)
        server_addr = server_sock.getsockname()[:2]
        process.start()
        async with await connect_tcp(*server_addr) as stream:
            await stream.send(b'GET')
            await stream.receive()

        process.join()
```

**Comment by agronholm** [comment]
OK, AnyIO v3.3.1 is out now.

**Comment by kangzhang** [comment]
Giving it a shot. Just upgraded to it. Thank you!

**Comment by kangzhang** [comment]
Had it running for a week and have not seen the exception since then. I think this is now fixed and we can close it. Thanks again for the help!

**Comment by tomchristie** [comment]
Thanks for the update @kangzhang - closing this off.
